aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePluginConfiguration.java

```diff
@@ -18,53 +18,134 @@
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 
-import com.amplifyframework.AmplifyException;
-
 import org.json.JSONException;
 import org.json.JSONObject;
 
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.concurrent.TimeUnit;
+
 /**
  * Configuration options for the {@link AWSDataStorePlugin}.
  * Contains settings for remote synchronization, if enabled.
  */
-final class AWSDataStorePluginConfiguration {
+public final class AWSDataStorePluginConfiguration {
+    /**
+     * At most one base sync may be performed within the period (now() - base sync interval).
+     * This interval has a default value provided, in case the user doesn't elect one.
+     */
+    public static final long DEFAULT_BASE_SYNC_INTERVAL_MS = TimeUnit.HOURS.toMillis(24);
+    private static final SyncMode DEFAULT_SYNC_MODE = SyncMode.LOCAL_ONLY;
 
     private final SyncMode syncMode;
+    private final long baseSyncIntervalMs;
 
-    private AWSDataStorePluginConfiguration(final SyncMode syncMode) {
+    private AWSDataStorePluginConfiguration(SyncMode syncMode, long baseSyncIntervalMs) {
         this.syncMode = syncMode;
+        this.baseSyncIntervalMs = baseSyncIntervalMs;
     }
 
     static AWSDataStorePluginConfiguration fromJson(JSONObject pluginJson) throws DataStoreException {
-        // If no sync mode is specified, we just use the default (no sync) and continue
-        if (pluginJson == null || !pluginJson.has("syncMode")) {
-            return new AWSDataStorePluginConfiguration(SyncMode.LOCAL_ONLY);
-        }
+        SyncMode syncMode = DEFAULT_SYNC_MODE;
+        long baseSyncIntervalMs = DEFAULT_BASE_SYNC_INTERVAL_MS;
 
-        try {
-            // If user has specified a sync mode, find out what it was.
-            final SyncMode syncMode = SyncMode.fromJsonPropertyValue(pluginJson.getString("syncMode"));
+        if (pluginJson == null) {
+            return new AWSDataStorePluginConfiguration(syncMode, baseSyncIntervalMs);
+        }
 
-            return new AWSDataStorePluginConfiguration(syncMode);
-        } catch (JSONException exception) {
-            throw new DataStoreException(
+        final Iterator<String> jsonKeys = pluginJson.keys();
+        while (jsonKeys.hasNext()) {
+            final String keyString = jsonKeys.next();
+            final ConfigKey configKey;
+            try {
+                configKey = ConfigKey.fromString(keyString);
+            } catch (IllegalArgumentException noSuchConfigKey) {
+                throw new DataStoreException(
```

Comment:
Hmm, it seems like this will throw an exception if it sees any key in the config it doesn't support, not just if it's missing the required keys. I think it would be better not to do this since (a) It doesn't hurt us in any way if there's extra information so why throw an error? and (b) I could see a time in the future where in transitioning between versions of our code in a non-breaking way there ends up being extraneous information in people's old config files which shouldn't be a problem but this would make into a problem.

Reply:
It also makes performance worse because you have to loop through the keys vs just check if it has the ones you want (not significant but still, given the above reasons, it adds fuel to the fire of removing it)

Reply:
Assuming I check all of the keys, the number of computations should be the same with or without the control flow.

I did intentionally code this to dis-allow unknown keys. The reasoning is that I don't want a user to have like a "syncMODE" and think it's okay. I want them to get a warning like "Hey bud, what's a syncMODE? The valid keys are: (acceptable keys including syncMode)."

I agree with the comments about transitioning keys in the future, but let's cross that bridge when we get to it. A different way of addressing it would be to have like:
```
// This is a 100% contrived, hypothetical example of us supporting keys during a graceful deprecation.
enum ConfigKey {
    /**
     * @deprecated Use {@link #SYNCHRONIZATION_MODE} instead.
     */
    @Deprecated
    SYNC_MODE("syncMode"),

    /**
     * How to synchronize data with the cloud?
     */
    SYNCHRONIZATION_MODE("syncrhonizationMode")
}
```
Then somewhere:
```
while (keyIterator.hasNext()) {
    ConfigKey configKey;
    try {
        configKey = ConfigKey.fromString(keyIterator.next());
    } catch (IllegalArgumentException noKnownConfigKey) {
        // Tell user what are the acceptable keys ....
    }
    switch (configKey) {
        case SYNC_MODE /* deprecated, but allowed for now... */:
        case SYNCHRONIZATION_MODE /* transitioning to this... */:
            // handle sync mode value
            break;
    }
}
```